tests/Aql/ExecutionBlockImplTest.cpp

```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
+                        ExecutionBlockImplExecuteSpecificTest, ::testing::Bool());
+
+enum CallAsserterState { INITIAL, SKIP, GET, COUNT, DONE };
+
+/**
+ * @brief Base class for call assertions.
+ * Every asserter holds an internal statemachine.
+ * And is called on every invocation of the LambdaFunction.
+ * According to it's internal machine, it asserts that the input Call
+ * Is expected in this situation.
+ *
+ */
+struct BaseCallAsserter {
+  // Actual number of calls for this machine
+  size_t call = 0;
+  // Maximum allowed calls to this machine, we assert that call <= maxCall
+  size_t maxCall = 0;
+  // Internal state
+  CallAsserterState state = CallAsserterState::DONE;
+  // The expected outer call, the machine needs to extract relevant parts
+  AqlCall const expected;
+
+  /**
+   * @brief Construct a new Base Call Asserter object
+   *
+   * @param expectedCall The given outer call. As we play several rounds (e.g. one call for skip one for get) the asserter needs to decompose this call
+   */
+  explicit BaseCallAsserter(AqlCall const& expectedCall)
+      : expected{expectedCall} {}
+
+  /**
+   * @brief Test if we need to expect a skip phase
+   *
+   * @return true Yes we have skip
+   * @return false No we do not have skip
+   */
+  auto hasSkip() const -> bool { return expected.getOffset() > 0; }
+  /**
+   * @brief Test if we need to expect a produce phase
+   *
+   * @return true
+   * @return false
+   */
+  auto hasLimit() const -> bool { return expected.getLimit() > 0; }
+  /**
+   * @brief Test if we need to expect a fullcount phase
+   *
+   * @return true
+   * @return false
+   */
+  auto needsFullCount() const -> bool { return expected.needsFullCount(); }
+};
+
+/**
+ * @brief Asserter used for the skipRows implementation.
+ *        Assumes that we are always called once with an empty input.
+ *        And once with a given input.
+ *        Will expect to be called for skip and fullCount (4 counts)
+ *        Does expect to not be called if skip and/or fullCount are ommited.
+ */
+struct SkipCallAsserter : public BaseCallAsserter {
+  bool firstCall = false;
+
+  explicit SkipCallAsserter(AqlCall const& expectedCall)
+      : BaseCallAsserter{expectedCall} {
+    // Calculate number of calls
+    // Ordering here is important, as it defines the start
+    // state of the asserter. We first get called for skip
+    // so skip needs to be last here
+    if (needsFullCount()) {
+      maxCall += 2;
+      state = CallAsserterState::COUNT;
+    }
+    if (hasSkip()) {
+      maxCall += 2;
+      state = CallAsserterState::SKIP;
+    }
+    // It is possible that we actually have 0 calls.
+    // if there is neither skip nor limit
+  }
+
+  auto gotCalled(AqlCall const& got) -> void {
+    call++;
+    switch (state) {
+      case CallAsserterState::SKIP: {
+        EXPECT_EQ(got.getOffset(), expected.getOffset());
+        firstCall = !firstCall;
+        if (!firstCall) {
```

Comment:
This is confusing. `firstCall` starts as `false`. And here you first invert it then test it. What does that mean? Also, it's done in every state.

As you already have `call`, I think it'd be more readable to just check `call % 2 == 1` or so.

Reply:
That is better indeed